torchao/quantization/quantize_/common/dynamic_activation_quantization_wrapper.py

```diff
@@ -0,0 +1,336 @@
+    """
+
+    def __new__(
+        cls,
+        original_weight_tensor: torch.Tensor,
+        act_tensor_kwargs: AOBaseTensorKwargs,
+    ):
+        kwargs = {}
+        dtype = original_weight_tensor.dtype
+        kwargs["dtype"] = dtype
+        kwargs["requires_grad"] = False
+        kwargs["device"] = original_weight_tensor.device
+        shape = original_weight_tensor.shape
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        original_weight_tensor: torch.Tensor,
+        act_tensor_kwargs: AOBaseTensorKwargs,
+    ):
+        self.original_weight_tensor = original_weight_tensor
+        self.act_tensor_kwargs = act_tensor_kwargs
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.act_tensor_kwargs}, {self.original_weight_tensor})"
+
+    def __tensor_flatten__(self):
+        return ["original_weight_tensor"], [self.act_tensor_kwargs]
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        original_weight_tensor = tensor_data_dict["original_weight_tensor"]
+        (act_tensor_kwargs,) = tensor_attributes
+        return cls(original_weight_tensor, act_tensor_kwargs)
+
+    @classmethod
+    def from_float(
+        cls,
+        weight: torch.Tensor,
+        act_tensor_kwargs: AOBaseTensorKwargs,
+    ):
+        return cls(weight, act_tensor_kwargs)
+
+    def _apply_fn_to_data(self, fn):
+        return self.__class__(
+            fn(self.original_weight_tensor),
+            self.act_tensor_kwargs,
+        )
+
+    def to(self, *args, **kwargs):
+        kwargs = self._get_to_kwargs(*args, **kwargs)
+        return self.__class__(
+            self.original_weight_tensor.to(**kwargs),
+            self.act_tensor_kwargs,
+        )
+
+
+def _same_metadata(
+    self: DynamicActivationQuantizationWrapper,
+    src: DynamicActivationQuantizationWrapper,
+):
+    return (
+        isinstance(self, DynamicActivationQuantizationWrapper)
+        and isinstance(src, DynamicActivationQuantizationWrapper)
+        and self.shape == src.shape
+        and self.act_tensor_kwargs == src.act_tensor_kwargs
+    )
+
+
+implements = DynamicActivationQuantizationWrapper.implements
+
+
+@implements([torch.nn.functional.linear, aten.linear.default])
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    if isinstance(weight_tensor, DynamicActivationQuantizationWrapper):
+        if input_tensor.numel() == 0:
+            return input_tensor
+        original_weight_tensor = weight_tensor.original_weight_tensor
+        kwargs = weight_tensor.act_tensor_kwargs
+        quantized_input = quantize_tensor(input_tensor, kwargs)
+        return torch.nn.functional.linear(quantized_input, original_weight_tensor, bias)
```

Comment:
IMO we want to call the linear override of `original_weight_tensor` here to keep things sane, not the override of `quantized_input`.  Can we clarify that this is working as intended?  If the override of `quantized_input` is being called, then it would be tricky to support multiple types of quantized activations with the same type of quantized weight.

If the override of `quantized_input` is being used, we can chat about how to make it less confusing, lmk